packages/tangle/solidifier.go

```diff
@@ -0,0 +1,246 @@
+		currentCachedMsgMetadata := currentSolidificationEntry.Value.([2]interface{})[1]
+		stack.Remove(currentSolidificationEntry)
+		return currentCachedMsg.(*CachedMessage), currentCachedMsgMetadata.(*CachedMessageMetadata)
+	}
+
+	// initialize the stack
+	solidificationStack := list.New()
+	solidificationStack.PushBack([2]interface{}{cachedMessage, cachedMsgMetadata})
+
+	// processed messages that are supposed to be checked for solidity recursively
+	for solidificationStack.Len() > 0 {
+		currentCachedMessage, currentCachedMsgMetadata := popElementsFromStack(solidificationStack)
+
+		currentMessage := currentCachedMessage.Unwrap()
+		currentMsgMetadata := currentCachedMsgMetadata.Unwrap()
+		if currentMessage == nil || currentMsgMetadata == nil {
+			currentCachedMessage.Release()
+			currentCachedMsgMetadata.Release()
+			continue
+		}
+
+		// check if the parents are solid
+		if t.isMessageSolid(currentMessage, currentMsgMetadata) {
+			// check parents age
+			valid := t.checkParentsBelowMaxDepth(currentMessage)
+			if !valid {
+				t.Events.MessageInvalid.Trigger(&CachedMessageEvent{
+					Message:         currentCachedMessage,
+					MessageMetadata: currentCachedMsgMetadata})
+				currentCachedMessage.Release()
+				currentCachedMsgMetadata.Release()
+				continue
+			}
+
+			// set the message solid
+			if currentMsgMetadata.SetSolid(true) {
+				t.Events.MessageSolid.Trigger(&CachedMessageEvent{
+					Message:         currentCachedMessage,
+					MessageMetadata: currentCachedMsgMetadata})
+
+				// auto. push approvers of the newly solid message to propagate solidification
+				t.Approvers(currentMessage.ID()).Consume(func(approver *Approver) {
+					approverMessageID := approver.ApproverMessageID()
+					solidificationStack.PushBack([2]interface{}{
+						t.Message(approverMessageID),
+						t.MessageMetadata(approverMessageID),
+					})
+				})
+			}
+		}
+
+		currentCachedMessage.Release()
+		currentCachedMsgMetadata.Release()
+	}
+}
+
+// deletes a message and its future cone of messages/approvers.
+// nolint
+func (t *Tangle) deleteFutureCone(messageID MessageID) {
+	cleanupStack := list.New()
+	cleanupStack.PushBack(messageID)
+
+	processedMessages := make(map[MessageID]types.Empty)
+	processedMessages[messageID] = types.Void
+
+	for cleanupStack.Len() >= 1 {
+		currentStackEntry := cleanupStack.Front()
+		currentMessageID := currentStackEntry.Value.(MessageID)
+		cleanupStack.Remove(currentStackEntry)
+
+		t.DeleteMessage(currentMessageID)
+
+		t.Approvers(currentMessageID).Consume(func(approver *Approver) {
+			approverID := approver.ApproverMessageID()
+			if _, messageProcessed := processedMessages[approverID]; !messageProcessed {
+				cleanupStack.PushBack(approverID)
+				processedMessages[approverID] = types.Void
+			}
+		})
+	}
+}
+
+// checks whether the given message is solid and marks it as missing if it isn't known.
+func (t *Tangle) checkParentsBelowMaxDepth(msg *Message) bool {
```

Comment:
BelowMaxDepth has a meaning only wrt to milestones.
```suggestion
func (t *Tangle) checkParentsMaxAge(msg *Message) bool {
```